storage: fix stopper race in compactor

tbg opened the pull request:
Starting workers without a surrounding task is unfortunately often not
the right thing to do when the worker accesses other state that might
become invalidated once the stopper begins to stop. In this particular
case, the compactor might end up accessing the engine even though it
had already been closed.

I wasn't able to repro this failure in the first place, but pretty sure
this:
Fixes #27232.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27699)
<!-- Reviewable:end -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27699#-:-LHhal3w1ycjDv917CAm:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

tschottdorf commented:
bors r=petermattis

On Wed, Jul 18, 2018 at 3:36 PM Peter Mattis <notifications@github.com>
wrote:

> *@petermattis* approved this pull request.
>
> [image: :lgtm:]
> <https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67>
>
> *Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/27699#-:-LHhal3w1ycjDv917CAm:bnfp4nl>*
> status: [image: :shipit:] complete! 0 of 0 LGTMs obtained (and 1 stale)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/27699#pullrequestreview-138256992>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135L4uTce839aav72PcOdFtE5a7baAks5uHznmgaJpZM4VUk7f>
> .
>
-- 

-- Tobias

tschottdorf commented:
bors r-

errcheck.

tbg committed:
storage: fix stopper race in compactor

Starting workers without a surrounding task is unfortunately often not
the right thing to do when the worker accesses other state that might
become invalidated once the stopper begins to stop. In this particular
case, the compactor might end up accessing the engine even though it
had already been closed.

I wasn't able to repro this failure in the first place, but pretty sure
this:
Fixes #27232.

Release note: None

tschottdorf commented:
bors r=petermattis

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27699)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/compactor/compactor.go, line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27699#-LHnN5PhCTGDYWOHRf9u:-LHnN5PhCTGDYWOHRf9v:bp4y14g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d182effd7a8e27c3f2eb7d3388bead17ccaf4703/pkg/storage/compactor/compactor.go#L112)):*
> ```Go
> 	// Run the Worker in a Task because the worker holds on to the engine and
> 	// may still access it even though the stopper has allowed it to close.
> 	_ = stopper.RunTask(ctx, "compactor", func(ctx context.Context) {
> ```

This seems backwards. It should be safe to start workers without a surrounding task at startup (which this is), but those workers should create internal tasks for the work that they do. The worker-in-task pattern is only needed for workers that are created after startup. 


<!-- Sent from Reviewable.io -->

tschottdorf commented:
That might be true, but the option you prefer also makes it much more
likely to mess this up again. Either way, I'm going to leave as is. It's
unfortunate that the API we came up with is so unfriendly.

On Thu, Jul 19, 2018 at 6:31 PM Ben Darnell <notifications@github.com>
wrote:

> *@bdarnell* commented on this pull request.
>
> *Reviewable <https://reviewable.io/reviews/cockroachdb/cockroach/27699>*
> status: [image: :shipit:] complete! 0 of 0 LGTMs obtained (and 1 stale)
>
> ------------------------------
>
> *pkg/storage/compactor/compactor.go, line 112 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/27699#-LHnN5PhCTGDYWOHRf9u:-LHnN5PhCTGDYWOHRf9v:bp4y14g>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/d182effd7a8e27c3f2eb7d3388bead17ccaf4703/pkg/storage/compactor/compactor.go#L112>):*
>
> 	// Run the Worker in a Task because the worker holds on to the engine and
> 	// may still access it even though the stopper has allowed it to close.
> 	_ = stopper.RunTask(ctx, "compactor", func(ctx context.Context) {
>
> This seems backwards. It should be safe to start workers without a
> surrounding task at startup (which this is), but those workers should
> create internal tasks for the work that they do. The worker-in-task pattern
> is only needed for workers that are created after startup.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/27699#pullrequestreview-138754284>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135Ego1XjiAeW1TwV58NOKM7RTkmiGks5uILRHgaJpZM4VUk7f>
> .
>
-- 

-- Tobias
